cordova-common/src/FileUpdater.js

```diff
@@ -0,0 +1,389 @@
+ *     path parameters are relative, or null if the paths are absolute. The rootDir is omitted
+ *     from any logged paths, to make the logs easier to read.
+ * @param {string} targetPath Destination file or directory to be updated. If it does not exist,
+ *     it will be created.
+ * @param {fs.Stats|null} targetStats An instance of fs.Stats for the target path, or null if
+ *     the target does not exist.
+ * @param {string|null} sourcePath Source file or directory to be used to update the
+ *     destination. If the source is null, then the destination is deleted if it exists.
+ * @param {fs.Stats|null} sourceStats An instance of fs.Stats for the source path, or null if
+ *     the source does not exist.
+ * @param {boolean} force If target and source are both files, and the force flag is not
+ *     set, then the file will not be copied unless the source is newer than the target.
+ * @param {function} [log] Optional logging callback that takes a string message describing any
+ *     file operations that are performed.
+ * @return {boolean} true if any changes were made, or false if the force flag is not set
+ *     and everything was up to date
+ */
+function updatePathWithStats(
+        rootDir, targetPath, targetStats, sourcePath, sourceStats, force, log) {
+    log = log || function(message) { };
+    var updated = false;
+
+    var targetFullPath = path.join(rootDir || "", targetPath);
+
+    if (sourceStats) {
+        var sourceFullPath = path.join(rootDir || "", sourcePath);
+
+        if (targetStats) {
+            // The target exists. But if the directory status doesn't match the source, delete it.
+            if (targetStats.isDirectory() && !sourceStats.isDirectory()) {
+                log("rmdir " + targetPath + " (source is a file)");
+                shell.rm("-rf", targetFullPath);
+                targetStats = null;
+                updated = true;
+            } else if (!targetStats.isDirectory() && sourceStats.isDirectory()) {
+                log("delete " + targetPath + " (source is a directory)");
+                shell.rm("-f", targetFullPath);
+                targetStats = null;
+                updated = true;
+            }
+        }
+
+        if (!targetStats) {
+            if (sourceStats.isDirectory()) {
+                // The target directory does not exist, so it should be created.
+                log("mkdir " + targetPath);
+                shell.mkdir("-p", targetFullPath);
+                updated = true;
+            } else if (sourceStats.isFile()) {
+                // The target file does not exist, so it should be copied from the source.
+                log("copy " + sourcePath + " " + targetPath +
+                    (!force ? " (new file)" : ""));
+                shell.cp("-f", sourceFullPath, targetFullPath);
+                updated = true;
+            }
+        } else if (sourceStats.isFile() && targetStats.isFile() &&
```

Comment:
The `if` block below is executed when `targetStats` is defined so probably id would be better to place it right after [line 72](https://github.com/apache/cordova-lib/pull/429/files#diff-8b56245d31f26df256f8361f42457c84R72)